README.md
# GTIRB GUI

A VSCode extension for viewing, navigating, and rewriting [GTIRB][], Grammatech's IR for Binaries.

[GTIRB]: https://github.com/GrammaTech/gtirb

## Quickstart
- Install gtirb-pprinter using instructions [installing][].
- Install pygls with pip, for example `pip install pygls`.
- Download the latest pre-built extension from the package registry and install it in VSCode (see "Installing the pre-built extension" below).
- Restart VSCode.
- Open a GTIRB file in VSCode which will now activate the extension.

[installing]: https://github.com/GrammaTech/gtirb/#installing

## Functionality

Some examples of what you can do with this extension: Load a GTIRB file and display an assembly file listing of the file contents. Hover to bring up any AuxData associated with the current line. Navigate to definitions by putting the cursor on a symbol and hitting F12 (or right click and select "Go To Definition").  You can also get referents to a function (or label) by going to the definition, right-clicking in the first block of instructions, and selecting "Find All References" (Alt+Shift+F12). More functionality will be added over time.

## Running in the repository

- Run `npm install` in this folder. This installs all necessary npm modules in both the client and server folder.
- Open VS Code on this folder.
- Open settings.json in the .vscode directory and check that "python.pythonPath" is correct for your system.
- Press Ctrl+Shift+B (Apple+Shift+B on a Mac) to compile the extension.
- Switch to the Debug viewlet.
- Select `Launch Client+Server` from the drop down.
- Run the launch config. This will start a TCP-based LSP server and add the client code to VSCode.
- In the [Extension Development Host] instance of vscode, open a folder with a ".gtirb" file in it and click the file to open it.

## Installing the pre-built extension

To install the extension as prebuilt VSIX file, go to the package registry and click on the latest version of gtirb-vscode. You will see a list of builds for this version, click on the most recent build to download it. To install into VSCode, run the following command in the folder you downloaded to:
```
% code --install-extension gtirb-vsix-VERSION.vsix
```
And start or restart vscode. The extension will start up automatically whenever you open a file with the ".gtirb" extension. You can uninstall the extension from within the GUI, or with the following command:
```
% code --uninstall-extension gtirb-vsix-VERSION.vsix
```

## Configuration options

The extension adds two configuration settings to configure the connection between client and LSP server. You can find these in the extension settings under "GTIRB Server Configuration"
- **Host**: (IP address of server): Set this to "localhost" (default) for an LSP server running on the same host as the client, or host IP address if remote. To configure the extension to create a dedicated LSP server running as a subprocess to the GTIRB extension (client), use "stdio" for host. The subprocess will exit when the client disconnects.
- **Port** (Port to connect to): This defaults to 3036, but can be changed to any convenient port number. If you are using something other than the default, be sure to use the same port when starting the LSP server. When the LSP server is a dedicated subprocess (host = "stdio") the port number is ignored.

## Installing the LSP server as a separate python package

To install with pip, got to the gtirb_lsp_server directory and run:
```
% pip3 install .
```

The LSP server optionally supports some limited binary rewriting capability. Some additional python packages are needed, to install these with pip go to the gtirb_lsp_server directory and run:
```
% pip3 install -r requirements-rewriting.txt
```
Or, to install the GTIRB LSP server and dependencies, including packages needed for rewriting capability:
```
% pip3 install .[rewriting]
```

## Running the LSP server

Server start-up and connection may be automatic or manual depending on the way you are running:
- If you have installed the extension and the host (in GTIRB Server Configration settings) is "stdio", a server is started for you when the client starts up, and stopped when the client exits.
- If you are running in the repository, use the "Launch Server (TCP)" configuration to start an LSP server before starting a client, or use "Launch Client+Server" to start both. The server will not exit when the client disconnects, you can use Ctrl-C in the terminal window to close it.
- You can also start the LSP Server manually using the command line below.

To start the LSP server:
```
% gtirb_lsp_server --tcp [ --host HOST ] [ --port PORT ] [ --verbose | --very-verbose ]
```
Where:
- HOST is the server IP address, the default is 127.0.0.1
- PORT is the TCP port, the default is 3036
- Default logging is to report errors, use verbose or very-verbose for additional levels of logging.

*NOTE* VSCode records user configuration settings in a file that persists, even when extensions are uninstalled. You may need to edit this file to reset the GTIRB Server host configuration. The path to this file depends on your host OS and version, some possibilities are:
- Linux: $HOME/.config/Code/User/settings/json
- Windows: C:\Users\%USERNAME%\AppData\Roaming\Code\User\Settings.json

## Prerequisites

This extension uses an LSP (language server protocol) server to add features to the VSCode UI. The LSP server is included with the extension.
- GTIRB UI (client) dependencies:
    - [vscode][].
    - [gtirb-pprinter][] version 1.8.2 or higher, which supports UI listing mode.
- LSP Server (python) dependencies:
    - python 3.6+.
    - The [gtirb][] python package.
    - Also the [pygls][] python package is required (can be installed with pip).
 - To support rewriting of GTIRB files (optional) the LSP server requires the following additional python packages
    - [gtirb-rewriting][]
    - [gtirb-functions][]
    - [mcasm][]

[vscode]: https://code.visualstudio.com/download
[gtirb]: https://github.com/GrammaTech/gtirb
[mcasm]: https://github.com/GrammaTech/mc-asm
[pygls]: https://github.com/openlawlibrary/pygls
[gtirb-pprinter]: https://github.com/GrammaTech/gtirb-pprinter
[gtirb-rewriting]: https://github.com/GrammaTech/gtirb-rewriting
[gtirb-functions]: https://github.com/GrammaTech/gtirb-functions

## Copyright and Acknowledgments

Copyright (C) 2022 GrammaTech, Inc.

This code is licensed under the GPLv3 license. See the LICENSE file in the project root for license terms.

This project is sponsored by the Office of Naval Research, One Liberty Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #N68335-17-C-0700. The content of the information does not necessarily reflect the position or policy of the Government and no official endorsement should be inferred.


CHANGELOG.md
# 0.0.4

- First public release

# 0.0.3

- Cleanup of output logging to GTIRB server channel on client
- Additional testing, including adding a mock server
- Move to more standard layout for lsp server and add console script

# 0.0.2

* Introduces support for remote operation of GTIRB LSP server.

# 0.0.1

* Initial version


src/extension.ts
// Copyright (C) 2022 GrammaTech, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//
// This project is sponsored by the Office of Naval Research, One
// Liberty Center, 875 N. Randolph Street, Arlington, VA 22203 under
// contract #N68335-17-C-0700.  The content of the information does
// not necessarily reflect the position or policy of the Government
// and no official endorsement should be inferred.
//"use strict";
import * as net from "net";
import * as path from "path";
import * as vscode from 'vscode';
import {
    LanguageClient,
    LanguageClientOptions,
    ServerOptions
} from 'vscode-languageclient/node';

import { GtirbEditorProvider } from './gtirbEditor';

import { getAddressAndJump, getPathForListing } from './customCommands';

import {promises as fs} from 'fs';
import * as url from 'url';

export let client: LanguageClient;

function getClientOptions(): LanguageClientOptions {
    return {
        // Register the server for GT assembly documents
        documentSelector: [
            { scheme: "file", language: "gtgas" },
            { scheme: "file", language: "gtmips" },
        ],
        outputChannelName: "Gtirb Language Server",
        synchronize: {
            // Notify the server about file changes to '.clientrc files contain in the workspace
            fileEvents: vscode.workspace.createFileSystemWatcher("**/.clientrc"),
        },
    };
}


function isStartedInDebugMode(): boolean {
    return process.env.VSCODE_DEBUG_MODE === "true";
}

function startLangServerTCP(port: number, hostAddr: string): LanguageClient {
    const serverOptions: ServerOptions = () => {
        return new Promise((resolve /*, reject */) => {
            const clientSocket = new net.Socket();
            clientSocket.connect(port, hostAddr, () => {
                resolve({
                    reader: clientSocket,
                    writer: clientSocket,
                });
            });
        });
    };

    return new LanguageClient(
        'gtirbServer',
        `tcp lang server (port ${port})`,
        serverOptions,
        getClientOptions()
    );
}

function startLangServer(
    command: string,
    args: string[],
    cwd: string
): LanguageClient {
    const serverOptions: ServerOptions = {
        args,
        command,
        options: { cwd },
    };

    return new LanguageClient(command, serverOptions, getClientOptions());
}


const FALLBACK_PYTHON = 'python';

/**
 * Derive python path from Python extension API
 * https://github.com/microsoft/vscode-python/wiki/AB-Experiments#pythondeprecatepythonpath
 * https://github.com/vscode-restructuredtext/vscode-restructuredtext/pull/224/files
 */
async function getPythonPath(): Promise<string> {
    let pythonPath;
    try {
        const extension = vscode.extensions.getExtension("ms-python.python");
        if (extension) {
            const usingNewInterpreterStorage = extension.packageJSON?.featureFlags?.usingNewInterpreterStorage;
            if (usingNewInterpreterStorage) {
                if (!extension.isActive) {
                    await extension.activate();
                }
                pythonPath = extension.exports.settings.getExecutionDetails().execCommand[0];
            }
        }
    } catch (error) {
        console.error('Error obtaining Python extension python path', error);
    }
    if (!pythonPath) {
        pythonPath = vscode.workspace.getConfiguration("python").get<string>("defaultInterpreterPath");
    }
    if (!pythonPath) {
        pythonPath = vscode.workspace.getConfiguration("python").get<string>("pythonPath");
    }
    if (!pythonPath) {
        pythonPath = FALLBACK_PYTHON;
    }
    return pythonPath;
}

interface GtirbPushParams {
    uri: string;
    content: string;
}


export async function activate(context: vscode.ExtensionContext) {
    const port = vscode.workspace.getConfiguration().get<number>('gtirb.server.port');
    const hostAddr = vscode.workspace.getConfiguration().get<string>('gtirb.server.host');

    const cwd = path.join(__dirname, "..");
    const pythonPath = await getPythonPath();

    if (!pythonPath) {
        throw new Error("`python.pythonPath` is not set");
    }

    if (hostAddr === 'stdio') {
        client = startLangServer(pythonPath, ["-m", "gtirb_lsp_server.gtirb_lsp_server"], cwd);
    } else {
        client = startLangServerTCP(port!, hostAddr!);
    }

    // Register request handlers for custom requests
    client.onReady().then(function (x: any) {
        // Response to a "Get GTIRB File" request from the server
        client.onRequest("gtirbGetGtirbFile", async (params: string) => {
            const p = url.fileURLToPath(new url.URL(params));
            const buf = await fs.readFile(p);
            const text = buf.toString('base64');
            return{params, languageId: '', version: 0, text: text};
        });
        // Response to a "Push GTIRB File" request from the server
        client.onRequest("gtirbPushGtirbFile", async (params: GtirbPushParams) => {
            const gtirb_uri = params.uri;
            const content = params.content;
            const buf = Buffer.from(content, 'base64');
            const p = url.fileURLToPath(new url.URL(gtirb_uri));
            fs.writeFile(p, buf).then(
                val => { client.outputChannel.appendLine(`GTIRB from server: ${params.uri} transfer complete.`); },
                err => { throw new Error(`failed to write gtirb file ${params.uri}: ${err}`); }
            );
            return{languageId: '', version: 0, text: "OK"};
        });
    });

    // Register our custom editor providers
    context.subscriptions.push(client.start());
    context.subscriptions.push(GtirbEditorProvider.register(context, pythonPath));

    context.subscriptions.push(
        vscode.commands.registerCommand('gtirb-vscode.goToAddress', () => {
            getAddressAndJump();
        }),
        vscode.commands.registerCommand('gtirb-vscode.getPathForListing', (gtirbFile, isa) =>
            getPathForListing(gtirbFile, isa)
        )
    );
}


export function deactivate(): Thenable<void> {
    //client.sendRequest("exit"); (needed?)
    return client ? client.stop() : Promise.resolve();
}
